### ISSUE TITLE
OS X privilege escalation due to bad error handling in Install.framework suid helper

### ISSUE THREAD

**Issue by GoogleCodeExporter** [asanso/google-security-research#314]
```
The private Install.framework has a few helper executables in 
/System/Library/PrivateFrameworks/Install.framework/Resources,
one of which is suid root:

-rwsr-sr-x   1 root  wheel   113K Oct  1  2014 runner

Taking a look at it we can see that it's vending an objective-c Distributed 
Object :)
[ 
https://developer.apple.com/library/mac/documentation/Cocoa/Conceptual/DistrObje
cts/DistrObjects.html ]

The main function immediately temporarily drops privs doing
  seteuid(getuid()); setegid(getgid());

then reads line from stdin. It passes this to NSConnection 
rootProxyForConnectionWithRegisteredName to lookup that
name in the DO namespace and create a proxy to connect to it via.

It then allocates an IFInstallRunner which in its init method vends itself 
using a name made up of its pid, time() and random()

It then calls the setRunnerConnectionName method on the proxy to tell it the 
IFInstallRunner's DO name so that whoever
ran the runner can connect to the IFInstallRunner.

The IFRunnerMessaging protocol tells us the methods and prototypes of the 
remote methods we can invoke on the IFInstallRunner.

Most of the methods begin with a call to processKey which will set the euid 
back to root if the process can provide a valid admin
authorization reference from authd (I'm not totally sure how that bit works 
yet, but it's not important for the bug.) Otherwise the euid
will remain equal to the uid and the methods (like movePath, touchPath etc) 
will only run with the privs of the user.

The methods then mostly end with a call to restoreUIDs which will drop back to 
euid==uid if we did temporarily regain root privs (with the auth ref.)

Not all methods we can invoke are like that though...

IFInstallRunner setExternalAuthorizationRef calls

  seteuid(0);setegid(0);

to regain root privs without requiring any auth. It then calls 
AuthorizationCreateFromExternalForm passing the bytes of an NSData we give it.

If that call doesn't return 0 then the error branch calls syslog with the 
string: "Fatal error: unable to internalize authorization reference."
but there's actually nothing fatal, it just returns from the method, whereas 
the success branch goes on to restore euid and egid, which means
that if we can get AuthorizationCreateFromExternalForm to fail then we can get 
the priv dropping-regaining state machine out-of-sync :)

Getting AuthorizationCreateFromExternalForm to fail is trivial, just provide a 
malformed auth_ref (like "AAAAAAAAAAAAAAAAAAA" )

Now the next method we invoke will run with euid 0 even without having the 
correct auth ref :)

This PoC first calls setBatonPath to point the baton executable path to a 
localhost bind-shell then triggers the bug
and calls runTaskSecurely which will create an NSTask and launch the bind-shell 
with euid 0 :) We can then just nc to it and get a root shell

tl;dr:
the error path in setExternalAuthorizationRef should either be fatal or drop 
privs!

Make sure you have the latest xcode installed and run the get_shell.sh script 
to build and run the PoC.
```

Original issue reported on code.google.com by `ianb...@google.com` on 1 Apr 2015 at 3:16

Attachments:
- [runner_sploit_redist.zip](https://storage.googleapis.com/google-code-attachments/google-security-research/issue-314/comment-0/runner_sploit_redist.zip)
